### PR TITLE
Port `equal` from THC to ATen (CUDA)

### LINKED ISSUE #24557: Migrate `equal` from the TH to Aten (CUDA)
Porting TH operators is essential for code simplicity and performance reasons.

Porting guides and Q&A are available in umbrella issue: #24507

Feel free to add @VitalyFedyunin as a reviewer to get a prioritized review.

cc @ezyang @gchanan @zou3519

### PR DESCRIPTION
**Pull request by Baranowski**
Fixes #24557

ASV benchmark:

```
import torch

sizes = [
    (10**6,),
    (1000, 1000),
    (10, 10),
    (1, 2, 3, 4, 5, 6, 7, 8, 9, 10),
]

class EqualTrue:
    params = range(len(sizes))

    def setup(self, n):
        dims = sizes[n]
        self.a = torch.rand(dims, device='cuda')
        self.b = self.a.clone()

    def time_equal(self, n):
        torch.equal(self.a, self.b)

class EqualFalse:
    params = range(len(sizes))

    def setup(self, n):
        dims = sizes[n]
        self.a = torch.rand(dims, device='cuda')
        self.b = torch.rand(dims, device='cuda')

    def time_equal(self, n):
        torch.equal(self.a, self.b)
```

Old results:
```
[ 75.00%] ··· equal.EqualFalse.time_equal
[ 75.00%] ··· ======== ============
               param1
              -------- ------------
                 0       67.7±7μs
                 1       74.0±2μs
                 2      24.4±0.1μs
                 3      135±0.2μs
              ======== ============

[100.00%] ··· equal.EqualTrue.time_equal
[100.00%] ··· ======== ============
               param1
              -------- ------------
                 0      59.8±0.2μs
                 1      59.9±0.3μs
                 2      25.0±0.5μs
                 3      136±0.2μs
              ======== ============
```

New results:
```
[ 75.00%] ··· equal.EqualFalse.time_equal
[ 75.00%] ··· ======== ============
               param1              
              -------- ------------
                 0      44.4±0.2μs 
                 1      44.5±0.4μs 
                 2      31.3±0.3μs 
                 3      96.6±0.5μs 
              ======== ============

[100.00%] ··· equal.EqualTrue.time_equal
[100.00%] ··· ======== ============
               param1              
              -------- ------------
                 0      44.2±0.2μs 
                 1      44.6±0.2μs 
                 2      30.8±0.3μs 
                 3      97.3±0.2μs 
              ======== ============
```

### CODE REVIEW

**Comment by Baranowski** [comment]
I don't see anything for `equal` in the `benchmarks/` folder. Any suggestions on what usage patterns I should benchmark for?

**Comment by Baranowski** [comment]
@VitalyFedyunin should I benchmark this change?

**Comment by VitalyFedyunin** [comment]
Hi! Yes, simple timeit benchmark with different tensor shapes(sizes) will suffice.

**Review by VitalyFedyunin — DISMISSED** [review]

**Comment by Baranowski** [comment]
I wasn't expecting this in such a simple change but I'm seeing a perf regression. I will need to dig into this.

File: aten/src/ATen/native/cuda/ReduceOpsKernel.cu [hunk 1]
```diff
@@ -482,4 +482,25 @@ REGISTER_DISPATCH(min_values_stub, &min_values_kernel_cuda);
...
+  at::NoNamesGuard guard;
+  TORCH_CHECK(self.device() == src.device(), "Cannot compare two tensors on "
+      "different devices. Got: ", self.device(), " and ", src.device());
+  if (self.sizes() != src.sizes()) {
+    return false;
+  }
+  if (self.numel() == 0) {
+    return true;
+  }
+  return THCudaBoolTensor_minall(
```
Comment (Baranowski) [inline aten/src/ATen/native/cuda/ReduceOpsKernel.cu:499]: I'm not sure that calling this function from ATen (as opposed to TH) is ok.

**Review by Baranowski — COMMENTED** [review]
@VitalyFedyunin this is ready for a review.

I have put the benchmark results and the benchmark itself in the PR description. Turns out that wrapping the value in a scalar_tensor and then extracting it from there was taking too long.

**Review by VitalyFedyunin — APPROVED** [review]
Code looks good, but you need to rebase

**Comment by Baranowski** [comment]
@VitalyFedyunin this is rebased and should be ready to be merged.

File: aten/src/ATen/native/cuda/ReduceLogicKernel.cu [hunk 2]
```diff
@@ -24,4 +26,24 @@ void or_kernel_cuda(TensorIterator& iter) {
...
+      "different devices. Got: ", self.device(), " and ", src.device());
+  if (self.sizes() != src.sizes()) {
+    return false;
+  }
+  if (self.numel() == 0) {
+    return true;
+  }
+  return THCudaBoolTensor_minall(
+      globalContext().getTHCState(),
+      at::native::eq(self, src).unsafeGetTensorImpl()
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/cuda/ReduceLogicKernel.cu:45]: It is safer if you call: `at::native::eq(self, src).min()`

Comment (Baranowski) [reply aten/src/ATen/native/cuda/ReduceLogicKernel.cu:45]: That leads to serious performance regression.

Comment (VitalyFedyunin) [reply aten/src/ATen/native/cuda/ReduceLogicKernel.cu:45]: Sorry for going circles but can you please check with `at::native::eq(self, src).all()` https://pytorch.org/docs/master/tensors.html#torch.BoolTensor.all

Comment (Baranowski) [reply aten/src/ATen/native/cuda/ReduceLogicKernel.cu]: I'm shocked. Not only is there no regression, it seems to be significantly faster now. (Updated the benchmarks in PR message)

File: aten/src/ATen/native/cuda/ReduceLogicKernel.cu [hunk 3]
```diff
@@ -4,6 +4,8 @@
 #include <ATen/native/SharedReduceOps.h>
 #include <ATen/native/ReduceOps.h>
 
+#include <THC/THCTensorMathReduce.cuh>
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/cuda/ReduceLogicKernel.cu:7]: Not required

Comment (Baranowski) [reply aten/src/ATen/native/cuda/ReduceLogicKernel.cu]: Fixed

**Comment by VitalyFedyunin** [comment]
Sorry this one got lost in notifications. Can you please rebase so I can land.

**Comment by Baranowski** [comment]
@VitalyFedyunin I have rebased again. This is ready for merge

**Comment by Baranowski** [comment]
Just a heads up: I won't have the time to work on this after two weeks from now. I think this should be ready to merge.

**Pull request closed without merging**